File: Grammars/Table/TableGrammarTable.vb
Imports System.Xml.Serialization
Imports System.Reflection
Imports NCalc
Imports Utility

Namespace Table

    Public Enum Actions
        Random
        [Loop]
        [Select]
    End Enum

    Public Class TableGrammarTable

#Region "Event Arguments"
        Public Class EvaluateTableParameterArgs
            Inherits System.EventArgs

            Public Sub New(ByVal name As String,
                           ByVal args As ParameterArgs)
                Me._name = name
                Me._args = args
            End Sub

            Private _name As String
            Private _args As ParameterArgs

            Public ReadOnly Property Name As String
                Get
                    Return _name
                End Get
            End Property
            Public ReadOnly Property Args As ParameterArgs
                Get
                    Return _args
                End Get
            End Property
        End Class

        Public Class EvaluateTableFunctionArgs
            Inherits System.EventArgs

            Public Sub New(ByVal name As String,
                           ByVal args As FunctionArgs)
                Me._name = name
                Me._args = args
            End Sub

            Private _name As String
            Private _args As FunctionArgs

            Public ReadOnly Property Name As String
                Get
                    Return _name
                End Get
            End Property
            Public ReadOnly Property Args As FunctionArgs
                Get
                    Return _args
                End Get
            End Property
        End Class
#End Region

#Region "Events"
        Public Event EvaluateParameter As EventHandler(Of EvaluateTableParameterArgs)
        Public Event EvaluateFunction As EventHandler(Of EvaluateTableFunctionArgs)
#End Region

        Private Const COMMENT_TOKEN As String = "#"

        Private _table As DataTable
        Private WithEvents _ncalc As New Calculator()

        <XmlAttribute("action")>
        Public Property Action() As Actions = Actions.Random
        <XmlAttribute("column")>
        Public Property Column() As String
        <XmlAttribute("randomModifier")>
        Public Property RandomModifier() As String
        <XmlAttribute("loopId")>
        Public Property LoopId() As String
        <XmlAttribute("repeat")>
        Public Property Repeat As String
        <XmlAttribute("repeatJoin")>
        Public Property RepeatJoin As String
        <XmlAttribute("selectValue")>
        Public Property SelectValue As String
        <XmlAttribute("name")>
        Public Property Name() As String
        <XmlAttribute("delimiter")>
        Public Property Delimiter() As String
        <XmlAttribute("skipTable")>
        Public Property SkipTable() As String
        <XmlText()>
        Public Property Value() As String
        <XmlIgnore()>
        Public ReadOnly Property Table As DataTable
            Get
                Return _table
            End Get
        End Property

        Public Function ProcessTable() As Dictionary(Of String, Object)
            If Not String.IsNullOrWhiteSpace(SkipTable) AndAlso CBool(_ncalc.Evaluate(SkipTable)) Then Return New Dictionary(Of String, Object)
            Select Case Action
                Case Actions.Random : Return ProcessRandom()
                Case Actions.Loop : Return ProcessLoop()
                Case Actions.Select : Return ProcessSelect()
            End Select
            Return New Dictionary(Of String, Object)
        End Function

        Private Function ProcessSelect() As Dictionary(Of String, Object)
            Dim value As String = SelectValue
            Dim row As DataRow
            Dim result As New Dictionary(Of String, Object)
            If value.StartsWith("=") Then
                value = _ncalc.Evaluate(value.Substring(1)).ToString
            End If

            ParseTable()

            row = _table.AsEnumerable.FirstOrDefault(Function(r As DataRow) r.Item(Column).Equals(value))

            ProcessRow(result, row)
            Return result
        End Function

        Private Function ProcessLoop() As Dictionary(Of String, Object)
            Dim result As New Dictionary(Of String, Object)
            Dim count As Int32 = GetRepeat()

            ParseTable()

            For i As Int32 = 1 To count
                For Each row As DataRow In _table.Rows
                    Dim id As String = row(LoopId).ToString
                    For Each column As DataColumn In Table.Columns
                        Dim key As String = id & "." & column.ColumnName
                        Dim expression As String = row(column.ColumnName).ToString
                        Dim value As Object
                        If expression.StartsWith("=") Then
                            value = _ncalc.Evaluate(expression.Substring(1))
                        Else
                            value = expression
                        End If
                        If result.ContainsKey(key) Then
                            result(key) = result(key).ToString & RepeatJoin & value.ToString
                        Else
                            result.Add(key, value)
                        End If
                    Next
                Next
            Next

            Return result
        End Function

        Private Function ProcessRandom() As Dictionary(Of String, Object)
            If _table Is Nothing Then ParseTable()
            Dim random As New Random
            Dim max As Int32 = CInt(_table.AsEnumerable.Max(Function(r As DataRow) CInt(r(Column))))
            Dim value As Int32
            Dim result As New Dictionary(Of String, Object)
            Dim modifier As Int32 = 0
            Dim count As Int32 = GetRepeat()

            For i As Int32 = 1 To count
                Dim selectedRow As DataRow = Nothing
                Dim index As Int32 = 0

                If Not String.IsNullOrWhiteSpace(RandomModifier) Then
                    If RandomModifier.StartsWith("=") Then
                        modifier = CInt(_ncalc.Evaluate(RandomModifier.Substring(1)))
                    Else
                        modifier = CInt(_ncalc.Evaluate(RandomModifier))
                    End If
                End If

                value = BaseGrammar.Random.Next(Math.Abs(max + modifier)) * If(max + modifier < 0, -1, 1)
                Do While index < Table.Rows.Count AndAlso selectedRow Is Nothing
                    Dim row As DataRow = Table.Rows(index)
                    If value < CInt(row(Column)) Then selectedRow = row
                    index += 1
                Loop

                If selectedRow Is Nothing Then selectedRow = Table.Rows(Table.Rows.Count - 1)

                ProcessRow(result, selectedRow)
            Next
            Return result
        End Function

        Private Sub ProcessRow(ByVal result As Dictionary(Of String, Object),
                               ByVal row As DataRow)
            If row IsNot Nothing Then
                For Each column As DataColumn In row.Table.Columns
                    Dim thisValue As Object
                    If row(column).ToString.StartsWith("=") Then
                        thisValue = _ncalc.Evaluate(row(column).ToString.Substring(1))
                    Else
                        thisValue = row(column)
                    End If
                    If result.ContainsKey(column.ColumnName) Then
                        result(column.ColumnName) = result(column.ColumnName).ToString & RepeatJoin & thisValue.ToString
                    Else
                        result.Add(column.ColumnName, thisValue)
                    End If
                Next
            End If
        End Sub

        Private Function GetRepeat() As Int32
            If String.IsNullOrWhiteSpace(Repeat) Then Return 1
            If Repeat.StartsWith("=") Then
                Return CInt(_ncalc.Evaluate(Repeat.Substring(1)))
            Else
                Return CInt(Repeat)
            End If
        End Function

        Private Sub ParseTable()
            If _table Is Nothing Then
                Dim reader As New IO.StringReader(Value)
                Dim parser As New Microsoft.VisualBasic.FileIO.TextFieldParser(reader)
                parser.Delimiters = {Delimiter}
                parser.TextFieldType = FileIO.FieldType.Delimited
                parser.CommentTokens = {COMMENT_TOKEN}
                parser.HasFieldsEnclosedInQuotes = False
                parser.TrimWhiteSpace = True
                _table = New DataTable
                _table.TableName = Name

                ' Read headers
                Dim headers() As String = parser.ReadFields
                For Each header As String In headers
                    If Not String.IsNullOrWhiteSpace(header) Then _table.Columns.Add(header)
                Next

                ' Read data
                Do While Not parser.EndOfData
                    Dim fields() As String = parser.ReadFields
                    _table.Rows.Add(fields)
                Loop
            End If
        End Sub

        Private Sub _ncalc_EvaluateFunction(name As String, args As FunctionArgs) Handles _ncalc.EvaluateFunction
            RaiseEvent EvaluateFunction(Me, New EvaluateTableFunctionArgs(name, args))
        End Sub

        Private Sub _ncalc_EvaluateParameter(name As String, args As ParameterArgs) Handles _ncalc.EvaluateParameter
            RaiseEvent EvaluateParameter(Me, New EvaluateTableParameterArgs(name, args))
        End Sub
    End Class
End Namespace

File: Grammars/Table/TableGrammarTable.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Xml.Serialization;
using Microsoft.VisualBasic.FileIO;
using NCalc;
using YamlDotNet.Serialization;

namespace Grammars.Table
{
    public enum Actions
    {
        Random,
        @Loop,
        @Select
    }

    public class TableGrammarTable
    {
        private const string COMMENT_TOKEN = "#";
        private readonly Calculator _ncalc = new Calculator();

        private EventHandler<EvaluateTableFunctionArgs> EvaluateFunctionEvent;


        private EventHandler<EvaluateTableParameterArgs> EvaluateParameterEvent;

        [XmlAttribute("action")]
        public Actions Action { get; set; } = Actions.Random;

        [XmlAttribute("column")]
        public string Column { get; set; }

        [XmlAttribute("randomModifier")]
        public string RandomModifier { get; set; }

        [XmlAttribute("loopId")]
        public string LoopId { get; set; }

        [XmlAttribute("repeat")]
        public string Repeat { get; set; }

        [XmlAttribute("repeatJoin")]
        public string RepeatJoin { get; set; }

        [XmlAttribute("selectValue")]
        public string SelectValue { get; set; }

        [XmlAttribute("name")]
        public string Name { get; set; }

        [XmlAttribute("delimiter")]
        public string Delimiter { get; set; }

        [XmlAttribute("skipTable")]
        public string SkipTable { get; set; }

        [XmlText]
        public string Value { get; set; }

        [XmlIgnore]
        [YamlIgnore]
        public DataTable Table { get; private set; }

        public Dictionary<string, object> ProcessTable()
        {
            if (!string.IsNullOrWhiteSpace(SkipTable) && Convert.ToBoolean(_ncalc.Evaluate(SkipTable)))
            {
                return new Dictionary<string, object>();
            }
            switch (Action)
            {
                case Actions.Random:
                    return ProcessRandom();
                case Actions.Loop:
                    return ProcessLoop();
                case Actions.Select:
                    return ProcessSelect();
            }
            return new Dictionary<string, object>();
        }

        private Dictionary<string, object> ProcessSelect()
        {
            var value = SelectValue;
            var result = new Dictionary<string, object>();

            if (value.StartsWith("="))
            {
                value = _ncalc.Evaluate(value.Substring(1)).ToString();
            }

            ParseTable();

            var row = Table.AsEnumerable().FirstOrDefault(r => r[Column].Equals(value));

            ProcessRow(result, row);

            return result;
        }

        private Dictionary<string, object> ProcessLoop()
        {
            var result = new Dictionary<string, object>();
            var count = GetRepeat();

            ParseTable();

            for (var i = 1; i <= count; i++)
            {
                foreach (DataRow row in Table.Rows)
                {
                    var id = row[LoopId].ToString();
                    foreach (DataColumn column in Table.Columns)
                    {
                        var key = id + "." + column.ColumnName;
                        var expression = row[column.ColumnName].ToString();
                        object value;
                        if (expression.StartsWith("="))
                        {
                            value = _ncalc.Evaluate(expression.Substring(1));
                        }
                        else
                        {
                            value = expression;
                        }
                        if (result.ContainsKey(key))
                        {
                            result[key] = result[key] + RepeatJoin + value;
                        }
                        else
                        {
                            result.Add(key, value);
                        }
                    }
                }
            }

            return result;
        }

        private Dictionary<string, object> ProcessRandom()
        {
            if (Table == null)
            {
                ParseTable();
            }

            var max = Table.AsEnumerable().Max(r => Convert.ToInt32(r[Column]));
            var result = new Dictionary<string, object>();
            var modifier = 0;
            var count = GetRepeat();

            for (var i = 1; i <= count; i++)
            {
                DataRow selectedRow = null;
                var index = 0;

                if (!string.IsNullOrWhiteSpace(RandomModifier))
                {
                    if (RandomModifier.StartsWith("="))
                    {
                        modifier = Convert.ToInt32(_ncalc.Evaluate(RandomModifier.Substring(1)));
                    }
                    else
                    {
                        modifier = Convert.ToInt32(_ncalc.Evaluate(RandomModifier));
                    }
                }

                var value = BaseGrammar.Random.Next(Math.Abs(max + modifier))*(max + modifier < 0 ? -1 : 1);
                while (index < Table.Rows.Count && selectedRow == null)
                {
                    var row = Table.Rows[index];
                    if (value < Convert.ToInt32(row[Column]))
                    {
                        selectedRow = row;
                    }
                    index++;
                }

                if (selectedRow == null)
                {
                    selectedRow = Table.Rows[Table.Rows.Count - 1];
                }

                ProcessRow(result, selectedRow);
            }
            return result;
        }

        private void ProcessRow(Dictionary<string, object> result, DataRow row)
        {
            if (row != null)
            {
                foreach (DataColumn column in row.Table.Columns)
                {
                    object thisValue;
                    if (row[column].ToString().StartsWith("="))
                    {
                        thisValue = _ncalc.Evaluate(row[column].ToString().Substring(1));
                    }
                    else
                    {
                        thisValue = row[column];
                    }

                    if (result.ContainsKey(column.ColumnName))
                    {
                        result[column.ColumnName] = result[column.ColumnName] + RepeatJoin + thisValue;
                    }
                    else
                    {
                        result.Add(column.ColumnName, thisValue);
                    }
                }
            }
        }

        private int GetRepeat()
        {
            if (string.IsNullOrWhiteSpace(Repeat))
            {
                return 1;
            }
            if (Repeat.StartsWith("="))
            {
                return Convert.ToInt32(_ncalc.Evaluate(Repeat.Substring(1)));
            }
            return int.Parse(Repeat);
        }

        private void ParseTable()
        {
            if (Table == null)
            {
                var reader = new StringReader(Value);
                var parser = new TextFieldParser(reader)
                {
                    Delimiters = new[] {Delimiter},
                    TextFieldType = FieldType.Delimited,
                    CommentTokens = new[] {COMMENT_TOKEN},
                    HasFieldsEnclosedInQuotes = false,
                    TrimWhiteSpace = true
                };
                Table = new DataTable {TableName = Name};

                // Read headers
                var headers = parser.ReadFields();
                foreach (var header in headers)
                {
                    if (!string.IsNullOrWhiteSpace(header))
                    {
                        Table.Columns.Add(header);
                    }
                }

                // Read data
                while (!parser.EndOfData)
                {
                    var fields = parser.ReadFields();
                    Table.Rows.Add(fields);
                }
            }
        }

        private void _ncalc_EvaluateFunction(string name, FunctionArgs args)
        {
            if (EvaluateFunctionEvent != null)
            {
                EvaluateFunctionEvent(this, new EvaluateTableFunctionArgs(name, args));
            }
        }

        private void _ncalc_EvaluateParameter(string name, ParameterArgs args)
        {
            if (EvaluateParameterEvent != null)
            {
                EvaluateParameterEvent(this, new EvaluateTableParameterArgs(name, args));
            }
        }

        public event EventHandler<EvaluateTableParameterArgs> EvaluateParameter
        {
            add
            {
                EvaluateParameterEvent =
                    (EventHandler<EvaluateTableParameterArgs>) Delegate.Combine(EvaluateParameterEvent, value);
            }
            remove
            {
                EvaluateParameterEvent =
                    (EventHandler<EvaluateTableParameterArgs>) Delegate.Remove(EvaluateParameterEvent, value);
            }
        }

        public event EventHandler<EvaluateTableFunctionArgs> EvaluateFunction
        {
            add
            {
                EvaluateFunctionEvent =
                    (EventHandler<EvaluateTableFunctionArgs>) Delegate.Combine(EvaluateFunctionEvent, value);
            }
            remove
            {
                EvaluateFunctionEvent =
                    (EventHandler<EvaluateTableFunctionArgs>) Delegate.Remove(EvaluateFunctionEvent, value);
            }
        }


        public class EvaluateTableParameterArgs : EventArgs
        {
            public EvaluateTableParameterArgs(string name, ParameterArgs args)
            {
                Name = name;
                Args = args;
            }

            public string Name { get; }

            public ParameterArgs Args { get; }
        }

        public class EvaluateTableFunctionArgs : EventArgs
        {
            public EvaluateTableFunctionArgs(string name, FunctionArgs args)
            {
                Name = name;
                Args = args;
            }

            public string Name { get; }

            public FunctionArgs Args { get; }
        }
    }
}